File: server/common/socket_wrapper.py

"""
Sockets protocol:
    - 2 bytes que indican el tamanio del mensaje
    - mensaje

    Ejemplo:
        Cuando el server confirma la apuesta, le envia un "OK" al cliente
    escribiendo:
    02OK
"""

MSG_LEN_SIZE = 2


def read_msg(socket):
    # Primero trato de leer los primeros 2 bytes que me dicen el tamanio
    # del mensaje
    msg_len_b = b""
    while len(msg_len_b) < MSG_LEN_SIZE:
        bytes_leidos = socket.recv(MSG_LEN_SIZE - len(msg_len_b))
        if not bytes_leidos:
            return None
        msg_len_b += bytes_leidos

    msg_len = int.from_bytes(msg_len_b, byteorder="big")
    # Ahora leo el mensaje completo
    msg = b""
    while len(msg) < msg_len:
        bytes_leidos = socket.recv(msg_len - len(msg))
        if not bytes_leidos:
            return None
        msg += bytes_leidos
    return msg.decode('utf-8')


def write_msg(socket, msg):
    # Formateo el mensaje
    encoded_msg = msg.encode('utf-8') # con utf8 1 caracter = 1 byte
    msg_length = len(encoded_msg).to_bytes(MSG_LEN_SIZE, byteorder="big")
    full_msg = msg_length + encoded_msg
    # Lo mando envitand short-writes
    bytes_totales = 0
    while bytes_totales < len(full_msg):
        bytes_enviados = socket.send(full_msg[bytes_totales:])
        if bytes_enviados == 0:
            raise RuntimeError("Socket broken")
        bytes_totales += bytes_enviados

File: client/common/socket_wrapper.go

package common

import (
	"net"
	"encoding/binary"
	"fmt"
)


// Sockets protocol:
//     - 2 bytes que indican el tamanio del mensaje
//     - mensaje

//     Ejemplo:
//         Cuando el server confirma la apuesta, le envia un "OK" al cliente
//     escribiendo:
//     02OK


func SocketWriter(conn net.Conn, msg string) error {
	MSG_LEN_SIZE := 2
	// Formateo el mensaje
    encodedMsg := []byte(msg) 
	msgLength := make([]byte, MSG_LEN_SIZE)
	binary.BigEndian.PutUint16(msgLength, uint16(len(encodedMsg))) 
	fullMsg :=append(msgLength, encodedMsg...)

	// Lo mando envitando short write
	bytesTotales := 0
	for bytesTotales < len(fullMsg) {
		bytesEnviados, err := conn.Write(fullMsg[bytesTotales:])
		if err != nil {
			return err		
		}
		bytesTotales += bytesEnviados
	}

    return nil
}


func SocketReader(conn net.Conn) (string, error) {
	// Primero leo los primeros 2 bytes para
	// saber el tamanio del mensaje
	MSG_LEN_SIZE := 2

	// Leo los primeros 2 bytes
	msgLenB := make([]byte, MSG_LEN_SIZE)
	bytesLeidos := 0
	for bytesLeidos < MSG_LEN_SIZE {
		n, err := conn.Read(msgLenB[bytesLeidos:])
		if err != nil {
			fmt.Println("Error al leer el tamanio del mensaje")
			fmt.Println(err)
			fmt.Println("Bytes leidos: ", bytesLeidos)
			return "", err
		}
		bytesLeidos += n
		
	}
	msgLen := int(binary.BigEndian.Uint16(msgLenB))

	// Ahora leo el mensaje completo
	msg := make([]byte, msgLen)
	bytesLeidosMsg := 0
	for bytesLeidosMsg < msgLen {
		n, err := conn.Read(msg[bytesLeidosMsg:msgLen])
		if err != nil {
			return "", err
		}
		bytesLeidosMsg += n
	}
	return string(msg), nil
}
